tests/drivers/flash_nop_device/src/main.c

```diff
@@ -0,0 +1,204 @@
+	uint32_t val32 = 0, r_val32;
+	int rc;
+
+	for (off = 0; off < TEST_NOP_FLASH_SIZE; off += 4) {
+		rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET +
+					    off,
+				 &val32, sizeof(val32));
+		zassert_ok(rc, "flash_write (%d) should succeed at off 0x%x", rc,
+			   FLASH_NOP_DEVICE_BASE_OFFSET + off);
+		val32++;
+	}
+
+
+	for (off = 0; off < TEST_NOP_FLASH_SIZE; off += 4) {
+		rc = flash_read(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET +
+					   off,
+				&r_val32, sizeof(r_val32));
+		zassert_ok(rc, "flash_read should succeed");
+		zassert_equal(PATTERN8TO32BIT(FLASH_NOP_DEVICE_ERASE_VALUE), r_val32,
+			"flash byte at offset 0x%x has value 0x%08x, expected  0x%08x",
+			off, r_val32, PATTERN8TO32BIT(FLASH_NOP_DEVICE_ERASE_VALUE));
+	}
+}
+
+static void test_erase(void)
+{
+	int rc;
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET +
+			 FLASH_NOP_DEVICE_ERASE_UNIT,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+	zassert_ok(rc, "flash_erase should succeed");
+}
+
+static void test_out_of_bounds(void)
+{
+	int rc;
+	uint8_t data[8] = {0};
+
+	rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, TEST_NOP_FLASH_END,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, TEST_NOP_FLASH_END - 4,
+				 data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET -
+			 FLASH_NOP_DEVICE_ERASE_UNIT,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, TEST_NOP_FLASH_END,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET -
+			 FLASH_NOP_DEVICE_ERASE_UNIT * 2,
+			 FLASH_NOP_DEVICE_ERASE_UNIT * 2);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, TEST_NOP_FLASH_END -
+			 FLASH_NOP_DEVICE_ERASE_UNIT,
+			 FLASH_NOP_DEVICE_ERASE_UNIT * 2);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, TEST_NOP_FLASH_END,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, TEST_NOP_FLASH_END - 4, data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+}
+
+static void test_align(void)
+{
+#if FLASH_NOP_DEVICE_ERASE_UNIT > 1
```

Comment:
PROG_UNIT ?
ERASE_UNIT affects only erase part, right?

Reply:
Sorry, yes. You are right.